delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ResourcesPractice.scala

```diff
@@ -0,0 +1,151 @@
+    * @param source
+    *   the original json payload
+    * @param caller
+    *   the user performing the action
+    */
+  def generate(project: ProjectRef, schema: IdSegment, source: NexusSource)(implicit
+      caller: Caller
+  ): IO[ResourceRejection, DataResource]
+
+  /**
+    * Generates the resource and validate it against the provided schema
+    *
+    * @param project
+    *   the enclosing project
+    * @param schema
+    *   the schema to validate on
+    * @param source
+    *   the original json payload
+    * @param caller
+    *   the user performing the action
+    */
+  def generate(project: ProjectRef, schema: ResourceF[Schema], source: NexusSource)(implicit
+      caller: Caller
+  ): IO[ResourceRejection, DataResource]
+
+  /**
+    * Validates an existing resource.
+    *
+    * @param id
+    *   the identifier that will be expanded to the Iri of the resource
+    * @param project
+    *   the project reference where the resource belongs
+    * @param schemaOpt
+    *   the optional identifier that will be expanded to the schema reference to validate the resource. A None value
+    *   uses the currently available resource schema reference.
+    */
+  def validate(id: IdSegmentRef, project: ProjectRef, schemaOpt: Option[IdSegment])(implicit
+      caller: Caller
+  ): IO[ResourceRejection, ValidationResult]
+}
+
+object ResourcesPractice {
+  def apply(
+      fetchResource: (IdSegmentRef, ProjectRef) => IO[ResourceFetchRejection, DataResource],
+      validateResource: ValidateResource,
+      fetchContext: FetchContext[ProjectContextRejection],
+      contextResolution: ResolverContextResolution
+  )(implicit api: JsonLdApi, clock: Clock[UIO], uuidF: UUIDF): ResourcesPractice = new ResourcesPractice {
+
+    private val sourceParser = JsonLdSourceResolvingParser[ResourceRejection](contextResolution, uuidF)
+
+    override def generate(project: ProjectRef, schema: IdSegment, source: NexusSource)(implicit
+        caller: Caller
+    ): IO[ResourceRejection, DataResource] =
+      for {
+        projectContext <- fetchContext.onRead(project)
+        schemaRef      <- Resources.expandResourceRef(schema, projectContext)
+        jsonld         <- sourceParser(project, projectContext, source.value)
+        validation     <- validateResource(jsonld.iri, jsonld.expanded, schemaRef, project, caller)
+        res            <- toResourceF(project, jsonld, source, validation)
+      } yield res
+
+    override def generate(project: ProjectRef, schema: ResourceF[Schema], source: NexusSource)(implicit
+        caller: Caller
+    ): IO[ResourceRejection, DataResource] =
+      for {
+        projectContext <- fetchContext.onRead(project)
+        jsonld         <- sourceParser(project, projectContext, source.value)
+        validation     <- validateResource(jsonld.iri, jsonld.expanded, schema)
+        res            <- toResourceF(project, jsonld, source, validation)
+      } yield res
```

Comment:
It feels like this could be simplified

Reply:
I agree, we have to refactor this along with `ResourcesImpl` but I would postpone that after the migration to Cats-Effects as it will likely affect how we fetch project context for example.
Some of it may also be included in the evaluate method in the case of `ResourcesImpl`, it was done this way because with `akka-cluster`/`akka-persistence`, it was spawning an actor and we needed to define its id before the evaluation